#214 provider competitive selection

trdmitry opened the pull request:
Added new field CompetitiveSelection to Workshop, WorkshopDto, WorkshopCardDto

trdmitry committed:
add CompetitiveSelection

* add CompetitiveSelection to dto, model
* changed migration

trdmitry committed:
update Migration

trdmitry committed:
fix tests

i-rakityanskiy commented:
There is a condition in the story that if workshop has no competitive selection, then applications cannot have `AcceptedForSelection` status. I think that a validation should be added for this in scope of this story.

i-rakityanskiy commented on OutOfSchool/OutOfSchool.DataAccess/Models/Workshop.cs:
```diff
@@ -56,6 +56,9 @@ public class Workshop : IKeyedEntity<Guid>, IImageDependentEntity<Workshop>
     [Range(0, 18, ErrorMessage = "Max age should be a number from 0 to 18")]
     public int MaxAge { get; set; }
 
+    [Required(ErrorMessage = "CompetitiveSelection is required")]
+    public bool CompetitiveSelection { get; set; } = default;
```
What is the purpose of setting `default` as the initial value here?
```suggestion
    [Required(ErrorMessage = "CompetitiveSelection is required")]
    public bool CompetitiveSelection { get; set; }
```

trdmitry replied:
fixed.

i-rakityanskiy commented on OutOfSchool/OutOfSchool.WebApi/Models/Workshop/WorkshopCard.cs:
```diff
@@ -38,6 +38,9 @@ public class WorkshopCard
     [Range(0, 18, ErrorMessage = "Max age should be a number from 0 to 18")]
     public int MaxAge { get; set; }
 
+    [Required(ErrorMessage = "CompetitiveSelection is required")]
+    public bool CompetitiveSelection { get; set; } = default;
```
Same here.

trdmitry replied:
fixed.

i-rakityanskiy commented on OutOfSchool/OutOfSchool.WebApi/Models/Workshop/WorkshopDTO.cs:
```diff
@@ -60,6 +60,9 @@ public class WorkshopDTO : IValidatableObject
     [Range(0, 18, ErrorMessage = "Max age should be a number from 0 to 18")]
     public int MaxAge { get; set; }
 
+    [Required(ErrorMessage = "CompetitiveSelection is required")]
+    public bool CompetitiveSelection { get; set; } = default;
```
Same here.

trdmitry replied:
fixed.

trdmitry committed:
remove initializers

trdmitry committed:
fix change status for application

for competitive selection workshops user can only change the status from Pending to AcceptedForSelection

trdmitry commented:
> There is a condition in the story that if workshop has no competitive selection, then applications cannot have `AcceptedForSelection` status. I think that a validation should be added for this in scope of this story.

Added check status of applications with workshop that has flag CompetetiveSelection

i-rakityanskiy commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ApplicationController.cs:
```diff
@@ -453,6 +453,12 @@ private void UpdateStatus(ShortApplicationDto applicationDto, ApplicationDto app
             }
         }
 
+        if (application.Workshop.CompetitiveSelection && application.Status == ApplicationStatus.Pending &&
+            applicationDto.Status != ApplicationStatus.AcceptedForSelection)
```
What if we want to reject a pending application and our workshop has competitive selection? It looks like a valid action but it will cause an exception.
On the other hand there is still missing validation for workshops without competitive selection. It is still possible for them to set `ApplicationStatus` to `AcceptedForSelection`.

trdmitry replied:
> What if we want to reject a pending application and our workshop has competitive selection? It looks like a valid action but it will cause an exception.

I thought about this case and discussed with Mykita Babenko.  Provider can reject or accept the application after AcceptedForSelection status only.

We can discuss this case again with Mykita
Imho provider should be able to reject the application without competitive selection.

trdmitry replied:
> On the other hand there is still missing validation for workshops without competitive selection. It is still possible for them to set `ApplicationStatus` to `AcceptedForSelection`.

Added validation and testcase for workshops without competitive selection.

i-rakityanskiy commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/ApplicationController.cs:
```diff
@@ -453,6 +453,12 @@ private void UpdateStatus(ShortApplicationDto applicationDto, ApplicationDto app
             }
         }
 
+        if (application.Workshop.CompetitiveSelection && application.Status == ApplicationStatus.Pending &&
```
Please add tests for different scenarios.

trdmitry replied:
Tests added

trdmitry committed:
add Application status permissions

RBiyovskiy commented on OutOfSchool/OutOfSchool.WebApi/Common/StatusPermissions/StatusPermissions.cs:
```diff
@@ -0,0 +1,53 @@
+
+        var allowResult = AllowedStatuses.Any(p =>
+            (p.Role == role || p.Role == "all")
+            && (Convert.ToInt32(p.FromStatus) == Convert.ToInt32(from) || Convert.ToInt32(p.FromStatus) == 0)
```
You use generic method and than the convert method in body of them. Maybe will be better convert value to int before give it to method and do not use generic method

trdmitry replied:
if I change type to int I should convert enum to int when call the method CanChangeStatus 
![image](https://user-images.githubusercontent.com/3643368/184628615-1719bbbe-517a-499f-bb99-56e53fcaee13.png)

With geneneric this class can be used for any enums

RBiyovskiy reviewed (commented):
Please, fix tests.

trdmitry committed:
fix test, naming

trdmitry committed:
fix tests

ApplicationControllerTests generates random status for application. StatusPermissions doesn't allow to change from any to any status.

trdmitry commented on OutOfSchool/OutOfSchool.WebApi/Common/StatusPermissions/StatusPermissions.cs:
```diff
@@ -0,0 +1,53 @@
+
+    public bool CanChangeStatus(string role, T from, T to)
+    {
+        var denyResult = AllowedStatuses.Any(p =>
```
Renamed to statusPermissionsList

trdmitry commented:
> Please, fix tests.

fixed

trdmitry committed:
migration rebase

trdmitry committed:
reorder migration, fix tests

trdmitry committed:
Add subrole check

DmyMi commented:
Implemented by another PR
